### PR TITLE
feat: add handleError/errorMessage helpers to eliminate instanceof Er…

### LINKED ISSUE #4207: Eliminate `if (err instanceof Error)` pattern through helper function
**Is your feature request related to a problem? Please describe.**

When you have to handle an error in a TypeScript `try/catch` block, you need to verify that the parameter in the `catch` block is an `Error` before accessing any properties on it. If the value thrown is not `instanceof Error` we can just throw it as-is. We have this pattern in numerous places in the codebase.

**Describe the solution you'd like**

Implement a helper function to reduce duplicating this pattern.

**Describe alternatives you've considered**

Can continue adopting this pattern:

```typescript
try {
    // some action that can throw error
} catch (err) {
    if (err instanceof Error) {
        // transform or doing something special
    }
    throw err;
}
```
But this pattern can be easily de-duplicated

**Additional context**

Came up while reviewing code in [zowe/zowex](https://github.com/zowe/zowex)

### PR DESCRIPTION
**Pull request by Maanvi212006**
fixes #4207

## Proposed changes
-  Adds two utility helpers to `zowe-explorer-api` to eliminate the repeated `if (err instanceof Error)` pattern across the codebase:
- `handleError(err, callback)` — calls the callback only if `err` is an instance of `Error`, replacing repetitive type-guard blocks in try/catch
- `errorMessage(err)` — safely extracts the error message string, replacing inline `err instanceof Error ? err.message : String(err)` ternaries

Both are exported from `packages/zowe-explorer-api/src/utils/ErrorUtils.ts` and re-exported via the package index.

Note: Two occurrences of `err instanceof Error` were intentionally left unchanged — they include additional content checks (`err.message.includes(...)`) that go beyond type-guarding and cannot be safely replaced with the helper.

## Release Notes

Milestone: NA

Changelog: Added `handleError` and `errorMessage` utility functions to reduce duplicated error type-checking patterns across the codebase

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (non-breaking change which adds or improves functionality)
- [ ] Breaking change (a change that would cause existing functionality to not work as expected)
- [ ] Documentation (Markdown, README updates)
- [ ] Other (please specify above in "Proposed changes" section)

## Checklist

**General**

- [ ] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/zowe-explorer-vscode/wiki/Best-Practices:-Contributor-Guidance) wiki
- [ ] All PR dependencies have been merged and published (if applicable)
- [ ] A GIF or screenshot is included in the PR for visual changes
- [ ] The pre-publish command has been executed:
  - **v2 and below:** `yarn workspace vscode-extension-for-zowe vscode:prepublish`
  - **v3:** `pnpm --filter vscode-extension-for-zowe vscode:prepublish`
- [ ] New ZE APIs are tested with extender types that haven't adopted yet to determine breaking changes. Can use Zowe zFTP marketplace extension.

**Code coverage**

- [ ] There is coverage for the code that I have added
- [ ] I have added new test cases and they are passing
- [x] I have manually tested the changes

**Deployment**

- [ ] I have tested new functionality with the FTP extension and profile verifying no extender profile type breakages introduced
- [ ] I have added developer documentation (if applicable)
- [ ] Documentation should be added to Zowe Docs
  - If you're an outside contributor, please post in the [#zowe-doc Slack channel](https://openmainframeproject.slack.com/archives/CC961JYMQ) to coordinate documentation.
  - Otherwise, please check with the rest of the squad about any needed documentation before merging.
- [ ] These changes may need ported to the appropriate branches (list here):

## Further comments
NA

### CODE REVIEW

**Comment by Maanvi212006** [comment]
The CI failures appear to be pre-existing issues unrelated to this PR:
- zowe-explorer-ftp-extension build failure exists on main branch
- audit failures are dependency security issues

All tests in zowe-explorer and zowe-explorer-api packages pass locally. Happy to rebase if needed!

**Comment by JWaters02** [comment]
> The CI failures appear to be pre-existing issues unrelated to this PR:
> 
> * zowe-explorer-ftp-extension build failure exists on main branch
> * audit failures are dependency security issues
> 
> All tests in zowe-explorer and zowe-explorer-api packages pass locally. Happy to rebase if needed!

This is not the case, all CI/CD apart from audit-all are passing in main branch, so you have a failing build from this PR. Please run `pnpm run build` to find out why your build is failing.

File: packages/zowe-explorer-api/CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,7 @@ All notable changes to the "zowe-explorer-api" extension will be documented in t
 ## TBD Release
 
 ### New features and enhancements
+- Added `handleError` and `errorMessage` utility functions to eliminate repetitive `if (err instanceof Error)` patterns across the codebase. [#4207](https://github.com/zowe/zowe-explorer-vscode/issues/4207)
```
Comment (traeok) [inline packages/zowe-explorer-api/CHANGELOG.md:9]: I'm seeing this line repeated across multiple versions in the changelog. Can we remove the duplicates?

Also, please add a blank line between the header "New features and enhancements" and the new changelog entry.

File: packages/zowe-explorer/CHANGELOG.md [hunk 2]
```diff
@@ -5,6 +5,7 @@ All notable changes to the "vscode-extension-for-zowe" extension will be documen
 ## TBD Release
 
 ### New features and enhancements
+- Replaced repetitive `if (err instanceof Error)` patterns with `handleError` and `errorMessage` utility functions from `@zowe/zowe-explorer-api`. [#4207](https://github.com/zowe/zowe-explorer-vscode/issues/4207)
```
Comment (traeok) [inline packages/zowe-explorer/CHANGELOG.md:9]: Same suggestion as for the Zowe Explorer API changelog - remove duplicates and insert blank line between changelog header and entry.

File: packages/zowe-explorer/src/utils/AuthUtils.ts [hunk 3]
```diff
@@ -407,10 +408,10 @@ export class AuthUtils {
             this.updateNodeToolTip(sessionNode, profile);
             sessionNode.setSessionToChoice(commonApi.getSession());
         } catch (err) {
-            if (err instanceof Error) {
+            void handleError(err, async (error) => {
```
Comment (traeok) [inline packages/zowe-explorer/src/utils/AuthUtils.ts]: `async` is not necessary for the callback to `handleError` - we aren't awaiting any functions inside the callback.

File: packages/zowe-explorer/src/utils/AuthUtils.ts [hunk 4]
```diff
@@ -100,9 +101,9 @@ export class AuthUtils {
                     AuthHandler.disableSequentialRequests(profile);
                     return callbackValue;
                 } catch (err) {
-                    if (err instanceof Error) {
-                        ZoweLogger.error(err.message);
-                    }
+                    await handleError(err, async (error) => {
```
Comment (traeok) [inline packages/zowe-explorer/src/utils/AuthUtils.ts]: See comment about unnecessary `async` keyword on callback function

File: packages/zowe-explorer/src/utils/UriHandler.ts [hunk 5]
```diff
@@ -39,10 +39,10 @@ export class ZoweUriHandler implements UriHandler {
             .then(async (_entry) => {
                 await commands.executeCommand("vscode.open", parsedUri, { preview: false });
             })
-            .catch((err) => {
-                if (err instanceof Error) {
-                    ZoweLogger.error(`Failed to open external URL: ${err.message}`);
-                }
+            .catch(async (err) => {
+                await handleError(err, async (error) => {
```
Comment (traeok) [inline packages/zowe-explorer/src/utils/UriHandler.ts]: See comment about unnecessary `async` keyword on callback function

File: packages/zowe-explorer/src/utils/ProfilesUtils.ts [hunk 6]
```diff
@@ -555,7 +556,7 @@ export class ProfilesUtils {
                     })
                 );
             } catch (err) {
-                if (err instanceof Error) {
+                void handleError(err, async (error) => {
```
Comment (traeok) [inline packages/zowe-explorer/src/utils/ProfilesUtils.ts]: See comment about unnecessary `async` keyword on callback function

**Review by traeok — CHANGES_REQUESTED** [review]

**Comment by traeok** [comment]
Thanks for working on this @Maanvi212006 - left some comments about duplicated changelog entries and unnecessary `async` keywords.

Also, please bump the test coverage to meet the 90% minimum threshold 👍

**Comment by Maanvi212006** [comment]
Hi @traeok — I've addressed all the feedback:

- Removed async from all callbacks that don't use await
- Fixed both CHANGELOGs — removed duplicate entries and added blank lines
- Added unit tests for ErrorUtils.ts (16 tests, coverage now at 95.09%)

**Review by traeok — CHANGES_REQUESTED** [review]
I'm still seeing a bunch of additions/modifications in both CHANGELOG.md files (`packages/zowe-explorer`, `packages/zowe-explorer-api`)

Consider checking out the CHANGELOG from main:
`git checkout main -- packages/zowe-explorer/CHANGELOG.md packages/zowe-explorer-api/CHANGELOG.md`

And then only add one entry to the top of each changelog under "New features and enhancements"

**Comment by Maanvi212006** [comment]
> I'm still seeing a bunch of additions/modifications in both CHANGELOG.md files (`packages/zowe-explorer`, `packages/zowe-explorer-api`)
> 
> Consider checking out the CHANGELOG from main: `git checkout main -- packages/zowe-explorer/CHANGELOG.md packages/zowe-explorer-api/CHANGELOG.md`
> 
> And then only add one entry to the top of each changelog under "New features and enhancements"

I am sorry @traeok , I missed it. I have updated it!
Can you please check and let me know?

**Comment by traeok** [comment]
@Maanvi212006 changelogs look good now, can you please address the merge conflicts?
Once those are addressed, I plan to approve. Thank you

**Review by anaxceron — APPROVED** [review]

**Comment by traeok** [comment]
For future PRs, please avoid force-pushing the same commits that were already pushed to the branch.
This makes it more difficult to review because we can no longer see "what has changed" per GitHub's capabilities - all contents show as new edits after these changes were made.

**Comment by Maanvi212006** [comment]
> For future PRs, please avoid force-pushing the same commits that were already pushed to the branch. This makes it more difficult to review because we can no longer see "what has changed" per GitHub's capabilities - all contents show as new edits after these changes were made.

Sure, I will keep that in mind from next time!

**Review by traeok — CHANGES_REQUESTED** [review]
Seeing duplicate changelog entries again in ZE API's changelog.md.
Please resolve by 6/18 if possible. Thanks.

**Review by traeok — APPROVED** [review]
Patch coverage is only missing in areas/scenarios that we did not previously cover in catch blocks, rather than missing coverage in new code. So I think its close enough to accept. Thanks @Maanvi212006

File: packages/zowe-explorer/src/trees/dataset/DatasetTree.ts [hunk 7]
```diff
@@ -534,9 +535,9 @@ export class DatasetTree extends ZoweTreeProvider<IZoweDatasetTreeNode> implemen
                 profile,
             });
         } catch (err) {
-            if (err instanceof Error) {
-                ZoweLogger.warn(`Skipping creation of favorited profile. ${err.toString()}`);
-            }
+            handleError(err, (error) => {
```
Comment (zFernand0) [inline packages/zowe-explorer/src/trees/dataset/DatasetTree.ts:538]: Curious if the linter will complain if we don't `void` the "potential" resulting promise.
```
            void handleError(err, (error) => {
```

File: packages/zowe-explorer/src/utils/UriHandler.ts [hunk 8]
```diff
@@ -39,10 +39,10 @@ export class ZoweUriHandler implements UriHandler {
             .then(async (_entry) => {
                 await commands.executeCommand("vscode.open", parsedUri, { preview: false });
             })
-            .catch((err) => {
-                if (err instanceof Error) {
-                    ZoweLogger.error(`Failed to open external URL: ${err.message}`);
-                }
+            .catch(async (err) => {
+                await handleError(err, (error) => {
```
Comment (zFernand0) [inline packages/zowe-explorer/src/utils/UriHandler.ts:43]: Should we be trying to avoid combining `promise.then.catch` with `async/await` whenever possible?
```
            .catch((err) => {
                void handleError(err, (error) => {
```

**Review by zFernand0 — APPROVED** [review]
Changes LGTM! 😋 
---
Made some comments, but nothing that requires changes on this PR. We can discuss further during standups 🙏

**Comment by Maanvi212006** [comment]
> ## Changes LGTM! 😋
> Made some comments, but nothing that requires changes on this PR. We can discuss further during standups 🙏

Sounds good, happy to follow up if the team decides to make a change
